internal/config/cgroupmanager/systemdmanager.go

```diff
@@ -0,0 +1,183 @@
+	return moveConmonToSystemdCgroup(cid, cgroupParent, conmonCgroup, pid)
+}
+
+type Systemdv2Manager struct{}
+
+// Name returns the name of the cgroup manager (systemd)
+func (*Systemdv2Manager) Name() string {
+	return SystemdCgroupManager
+}
+
+// IsSystemd returns that it is a systemd cgroup manager
+func (*Systemdv2Manager) IsSystemd() bool {
+	return true
+}
+
+// GetContainerCgroupPath takes arguments sandbox parent cgroup and container ID and returns
+// the cgroup path for that containerID. If parentCgroup is empty, it
+// uses the default parent system.slice
+func (*Systemdv2Manager) GetContainerCgroupPath(sbParent, containerID string) string {
+	return getContainerCgroupPath(sbParent, containerID)
+}
+
+// systemd cgroups (v1 and v2) have no difference in creating the container cgroup path
+func getContainerCgroupPath(sbParent, containerID string) string {
+	parent := defaultSystemdParent
+	if sbParent != "" {
+		parent = sbParent
+	}
+	return parent + ":" + scopePrefix + ":" + containerID
+}
+
+// GetSandboxCgroupPath takes the sandbox parent, and sandbox ID. It
+// returns the cgroup parent, cgroup path, and error.
+// it also checks there is enough memory in the given cgroup (4mb is needed for the runtime)
+func (*Systemdv2Manager) GetSandboxCgroupPath(sbParent, sbID string) (string, string, error) {
+	return getSandboxCgroupPathForSystemd(sbParent, sbID, cgroupMemorySubsystemMountPathV2, "memory.max")
+}
+
+func getSandboxCgroupPathForSystemd(sbParent, sbID, memorySubsystemPath, memoryPath string) (string, string, error) {
+	if sbParent == "" {
+		return "", "", nil
+	}
+
+	if len(sbParent) <= 6 || !strings.HasSuffix(path.Base(sbParent), ".slice") {
+		return "", "", fmt.Errorf("cri-o configured with systemd cgroup manager, but did not receive slice as parent: %s", sbParent)
+	}
+
+	cgroupParent := convertCgroupFsNameToSystemd(sbParent)
+
+	if err := verifyCgroupHasEnoughMemory(cgroupParent, memorySubsystemPath, "memory.limit_in_bytes"); err != nil {
+		return "", "", err
+	}
+
+	cgPath := cgroupParent + ":" + scopePrefix + ":" + sbID
+
+	return cgroupParent, cgPath, nil
+}
+
+func verifyCgroupHasEnoughMemory(cgroupParent, memorySubsystemPath, memoryMaxFilename string) error {
+	// check memory limit is greater than the minimum memory limit of 4Mb
+	// expand the cgroup slice path
+	slicePath, err := systemd.ExpandSlice(cgroupParent)
+	if err != nil {
+		return errors.Wrapf(err, "error expanding systemd slice path for %q", cgroupParent)
+	}
+
+	// read in the memory limit from the memory.limit_in_bytes file
+	fileData, err := ioutil.ReadFile(filepath.Join(memorySubsystemPath, slicePath, memoryMaxFilename))
+	if err != nil {
+		if os.IsNotExist(err) {
+			logrus.Warnf("Failed to find %s for slice: %q", memoryMaxFilename, cgroupParent)
+			return nil
+		} else {
+			return errors.Wrapf(err, "error reading %s file for slice %q", memoryMaxFilename, cgroupParent)
```

Comment:
The error from ioutil.ReadFile already contains the file name and the fact that reading failed. Not sure if you add value by wrapping it.

Reply:
I like having the wrap so there's a unique string to grep against. I dropped the MemoryMaxFilename though, that's certainly redundant